pkg/error_code/error_code.go

```diff
@@ -322,12 +339,105 @@ type notFoundErr struct{ CodedError }
 // If the error is already an ErrorCode it will use that code
 // Otherwise it will use NotFoundCode which gives HTTP 404
 func NewNotFoundErr(err error) ErrorCode {
-	return internalErr{NewCodedError(err, NotFoundCode)}
+	return notFoundErr{NewCodedError(err, NotFoundCode)}
 }
 
 var _ ErrorCode = (*notFoundErr)(nil)     // assert implements interface
 var _ HasClientData = (*notFoundErr)(nil) // assert implements interface
 
+// HasOperation is an interface to retrieve the operation that occurred during an error.
+// The end goal is to be able to see a trace of operations in a distributed system to quickly have a good understanding of what occurred.
+// Inspiration is taken from upspin error handling: https://commandcenter.blogspot.com/2017/12/error-handling-in-upspin.html
+// The relationship to error codes is not one-to-one.
+// A given error code can be triggered by multiple different operations,
+// just as a given operation could result in multiple different error codes.
+//
+// GetOperation is defined, but generally the operation should be retrieved with Operation().
+// Operation() will check if a HasOperation interface exists.
+// As an alternative to defining this interface
+// you can use an existing wrapper (OpErrCode via AddOp) or embedding (EmbedOp) that has already defined it.
+type HasOperation interface {
+	GetOperation() string
+}
+
+// Operation will return an operation string if it exists.
+// It checks for the HasOperation interface.
+// Otherwise it will return the zero value (empty) string.
+func Operation(v interface{}) string {
+	var operation string
+	if hasOp, ok := v.(HasOperation); ok {
+		operation = hasOp.GetOperation()
+	}
+	return operation
+}
+
+// EmbedOp is designed to be embedded into your existing error structs.
+// It provides the HasOperation interface already, which can reduce your boilerplate.
+type EmbedOp struct{ Op string }
+
+// GetOperation satisfies the HasOperation interface
+func (e EmbedOp) GetOperation() string {
+	return e.Op
+}
+
+// MakeOp is just an alternative way to construct EmbedOp.
+func MakeOp(operation string) EmbedOp { return EmbedOp{Op: operation} }
```

Comment:
I did not find any use of `EmbedOp` or `MakeOp` in this PR. If it is not necessary, you can delete them since we already have the `OpErrCode` to build the errorCode that containing the Op field.

Reply:
Yes, I should delete `MakeOp`. I was still thinking of providing `EmbedOp` as a building block for the cases in which one would like to maintain the original error type rather than have it wrapped up in a new structure.